Enable multiple environments with different databases

adamfortuna opened the pull request:
### Goals

- Allow for all data currently used by the volunteer portal to be accessed via a local database
- Create a setup script to create and populate a local database to enable development without production database access

## Todo

### Finish relationship updates
- [x] Figure out how to create a new initiative with an array of `role`, `event`, or `initiative` objects that get saved in a single database transaction
- [x] Get the relationship working for initiatives
  - [x] Update tests to verify the model relationship, schema and API
  - [x] Fix duplicate relation bug in this test case: `ENV=test python -m pytest -k test_initiative_roles_scoped`
- [x] Get the relationship working for initiatives -> events
  - [x] update tests to verify the model relationship, schema and API
- [x] Get the relationship working for initiatives -> highlighted items
    - [x] update tests to verify the model relationship, schema and API
- [x] Add back and updated commented out tests
- [x] Get `PersonalDonationRequests` endpoint working

### Finish Postgres + MySQL Work
- [x] Abstract out all mysql/postgres specific column information into a setting that's dynamically pulled from `settings` based on the configured database engine according to `config.yaml`.

### Testing

- [x] Get testing working on GitHub

## Notes

The current approach in this branch is to move our models from Pydantic models to SQLAlchemy models. Pydantic models will still be used for our return types used in FastAPI. This is similar to the approach laid out in the [sample FastAPI codebase](https://github.com/tiangolo/full-stack-fastapi-postgresql/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/app/schemas/item.py). This will allow us to have a different Pydantic model for each API endpoint. For example, if we return a different data representation of an object in the LIST API endpoint as in the CREATE API endpoint for a model.

There are still a bunch of things to work out: validation, testing, verification that the dynamic database adapters work (that use postgres locally, and mysql bigtable on production) and more.

## Setup

Steps to get this branch working will eventually be in the readme. Until then, here's a brief intro:

Install Postgres locally. You'll need to have a `postgres` table in your local database -- which should always be created by default.

Next, make a copy of the `api/config.yaml.example` file to `api/config.yaml`. Make any changes to that copy based on your local development setup usernames, passwords and IP addresses.

With Postgres running, setup the API app like normal – either with virtualenv or conda. Install all dependencies.

run `python bootstrap.py` (or `make bootstrap`). This will create the local database, create the tables and populate them with sample data. After that you should be able to run the app like normal.

This should work for running tests and accessing the website. If you'd like to setup a different test database, you can set that up in the config.yaml file and bootstrap it using `ENV=test make bootstrap` to create it. Side note: `make test` uses whatever database you have configured in the `test` section of your config.

To run a single test case, like this one that's failing, you can run `ENV=test python -m pytest -k test_initiative_roles_scoped
`.

adamfortuna committed:
Updated README and app for issues that came up during onboarding

adamfortuna committed:
Moved declaration of donation_link_requests db to before its used

adamfortuna committed:
Setting up the API/python side to support dynamic databases and add a bootstrap script to create/populate databases.

adamfortuna committed:
Setting up the API/python side to support dynamic databases and add a bootstrap script to create/populate databases.

adamfortuna committed:
Update bootstrap script to insert initiatives, volunteer events and voluneteer openings

adamfortuna committed:
Update bootstrap script to insert initiatives, volunteer events and voluneteer openings

adamfortuna committed:
Update the API requests to use Pydantic models for return types

adamfortuna committed:
Update the API requests to use Pydantic models for return types

adamfortuna committed:
Allow volunteer roles to return data via FastAPI and pydantic models

adamfortuna committed:
Allow volunteer roles to return data via FastAPI and pydantic models

adamfortuna committed:
Adding in new tests for sqlalchemy

adamfortuna committed:
Adding in new tests for sqlalchemy

adamfortuna committed:
Fix VolunteerRole ID lookup API endpoint filtering

adamfortuna committed:
Fix VolunteerRole ID lookup API endpoint filtering

adamfortuna committed:
Added volunteer events API endpoints

adamfortuna committed:
Added volunteer events API endpoints

adamfortuna committed:
Added initiative model, schema and endpoint - without related models

adamfortuna committed:
Added initiative model, schema and endpoint - without related models

adamfortuna committed:
Updated config for generators

adamfortuna committed:
Updated config for generators

adamfortuna committed:
Added initiaitive -> role relationship and tests

adamfortuna committed:
Added initiaitive -> role relationship and tests

adamfortuna committed:
Uncomment db creation script

adamfortuna committed:
Uncomment db creation script

adamfortuna committed:
Limit sqlalchemy imports to only those that are used

adamfortuna committed:
Limit sqlalchemy imports to only those that are used

jtuckerk commented on api/settings.py:
```diff
@@ -0,0 +1,50 @@
+
+# Generate Database URLs based on the DB connection information
+def generate_hf_mysql_db_address(connection) -> str:
+    secret_path = f'projects/humanity-forward/secrets/{db_secret_key}/versions/latest'
+    db_pass = secret_client.access_secret_version(request={"name": secret_path}).payload.data.decode('UTF-8')
```
It'd be nice if this could be skipped if a password is supplied so this could work fully offline/without any hf GCP permissions.
I know this isn't done so don't worry about responding, but this probably won't be the last comment on incomplete code.

jtuckerk replied:
ah I think I misunderstood this on the first go around. There was no generator supplied in the yaml so it was breaking there but I see now I can specify the url generator to use either local or remote in the yaml

jtuckerk commented on api/models/volunteer_event.py:
```diff
@@ -0,0 +1,32 @@
+
+    def __repr__(self):
+        return "<VolunteerEvent(event_uuid='%s', event_external_id='%s', name='%s')>" % (
+                                self.event_uuid, self.event_external_id, self.name)
```
I like this organization quite a bit

jtuckerk commented on api/api.py:
```diff
@@ -1,82 +1,55 @@
-@app.get("/api/volunteer_roles/", response_model=List[VolunteerRole])
-def get_all_volunteer_roles(fake_data: bool = False) -> List[VolunteerRole]:
-    return roles_dataset.get_linked_model_objects() if not fake_data else generate_fake_volunteer_roles_list()
+@app.get("/api/volunteer_roles/", response_model=List[VolunteerRoleSchema])
+def get_all_volunteer_roles(db: Session = Depends(get_db)) -> List[VolunteerRoleSchema]:
+    return db.query(VolunteerRole).all()
```
so the sqlalchemy base models are autoconverted to pydantic models? that's pretty neat!

I'm wondering if we could/would want to try to combine the schemas and models so we only have a single model/schema. I'm concerned about the complexity of this as a volunteer project. It might be a more complicated base class, but I'm hoping it'd simplify additions and debugging. I'm imagining having a mismatch somewhere in the schema->model->db chain could be really annoying to debug.

with this setup I'm imagining we could still have additional schemas defined for endpoints from which we only want to serve a subset of fields.

jtuckerk replied:
> I added a tweak that'll let us autoload Postgres for our needs using docker-compose, it's in the `/postgres` folder. One can just run `docker-compose up -d` in that folder (assuming docker-compose is [installed](https://docs.docker.com/compose/install/)) and it'll start up a containerized Postgres instance. The existing bootstrap approach works just fine for loading the test data, though we may want to look at getting a fully pre-initialized container with the test data already in it, ideally.

Nice! I like moving towards dockerizing everything

How easy is it to define the test data to pre-initialize the db with? I like the set up now since, to me, editiing the python to add rows or edit the schema is intuitive, and also might means if people want to use fake data for testing theyve already written those util functions

adamfortuna replied:
> so the sqlalchemy base models are autoconverted to pydantic models? that's pretty neat!

Yep! It duck types the passed in object as whatever the schema is and converts it. I was surprised when it worked to be honest haha.

> I'm wondering if we could/would want to try to combine the schemas and models so we only have a single model/schema. I'm concerned about the complexity of this as a volunteer project. It might be a more complicated base class, but I'm hoping it'd simplify additions and debugging. I'm imagining having a mismatch somewhere in the schema->model->db chain could be really annoying to debug.

That is a concern for sure. There are some libraries that auto-convert from a model to a schema like [pydantic-sqlalchemy](https://github.com/tiangolo/pydantic-sqlalchemy). Those would convert everything, then we'd overwrite everything that's renamed for the front-end.

> with this setup I'm imagining we could still have additional schemas defined for endpoints from which we only want to serve a subset of fields.

adamfortuna committed:
Commenting out UUID columns for now since MySQL doesn't have those

adamfortuna committed:
Commenting out UUID columns for now since MySQL doesn't have those

Calvinxc1 committed:
Added postgres docker scaffolding

Calvinxc1 committed:
Added postgres docker scaffolding

Calvinxc1 commented:
I added a tweak that'll let us autoload Postgres for our needs using docker-compose, it's in the `/postgres` folder. One can just run `docker-compose up -d` in that folder (assuming docker-compose is [installed](https://docs.docker.com/compose/install/)) and it'll start up a containerized Postgres instance. The existing bootstrap approach works just fine for loading the test data, though we may want to look at getting a fully pre-initialized container with the test data already in it, ideally.

Calvinxc1 committed:
Removed jupyter cache dir

Calvinxc1 committed:
Removed jupyter cache dir

jtuckerk committed:
Single DB config & psql boostrap setup

Reducing the config file and the bootstrap logic to expect only a
  single database. introducing the second database was only intended
  to be temporary to ensure no pii was accidentally shared.

  setting up the bootstrap script to match and starting the transition
  to psql code

jtuckerk committed:
in progress

jtuckerk committed:
Fix filter argument error - api with local db working

jtuckerk committed:
Add events relationships to initiatives.

jtuckerk commented:
I've rebased this change onto the development branch.

I haven't had the time I've wanted to get this fully spruced up to merge into develop. I think it is getting close to at least unblock people for more convenient local development.

The main piece that I think needs to be finished before we can merge this into develop is the tests. Im uncertain about some of the modules are set up and the dependencies referenced. I mostly have used python with bazel to manage dependencies and feel a bit out of my depth on this. @Calvinxc1 or @trippersham if either of you have opinions about how this should be done, please chime in

jtuckerk committed:
Adding donation link request model + minor fixes

jtuckerk committed:
Fix up tests

jtuckerk committed:
add end to end post link request test

jtuckerk committed:
remove base __init__ doesnt seem needed and messes with mypy

jtuckerk committed:
fix linting warnings

adamfortuna commented:
> I added a tweak that'll let us autoload Postgres for our needs using docker-compose, it's in the `/postgres` folder. One can just run `docker-compose up -d` in that folder (assuming docker-compose is [installed](https://docs.docker.com/compose/install/)) and it'll start up a containerized Postgres instance. The existing bootstrap approach works just fine for loading the test data, though we may want to look at getting a fully pre-initialized container with the test data already in it, ideally.

Ohh nice! I'm looking forward to giving that a try.

adamfortuna committed:
Add additional test to make sure initiatives relationships are scoped

adamfortuna committed:
Adding test for verifying relationships are loading right

adamfortuna committed:
Move test data into test directory

adamfortuna committed:
Changed config example to match current schema used

Calvinxc1 committed:
merge so I can progress

Calvinxc1 committed:
Added docker-compose for postgres, with test data

Calvinxc1 commented:
Ok, just merged a new docker setup. You should be able to do a `docker-compose build` followed by a `docker-compose up -d` to run the Postgres database. It'll be accessible on the same port that it is by default (5432). Also adminer is running at port 8080 so you have a GUI interface to explore the database. Enjoy!

adamfortuna committed:
Undoing the merge commit. Checked with Jason and looks like the Docker work is in the commit after this one.

This reverts commit 2e98b03395f49bf495b9a20832f619e8d7d6c909, reversing
changes made to 653b17c1e806fd82b81a33ba9886de29cf020883.

adamfortuna committed:
Switched from using postgres data files to use postgres schema dumps

adamfortuna committed:
WIP for the app docker setup:

adamfortuna committed:
Client application working with Docker

adamfortuna committed:
Uncommented the docker db setup since it's working

Calvinxc1 commented:
Getting the following when I try to run `docker-compose build` on the new code:
```
Step 6/6 : RUN npm run build
 ---> Running in e17d1234133e

> volunteer-portal@0.5.0 build /srv/client
> react-scripts build

Creating an optimized production build...
Failed to compile.

./src/styles/base.scss
Error: Missing binding /srv/client/node_modules/node-sass/vendor/linux-x64-72/binding.node
Node Sass could not find a binding for your current environment: Linux 64-bit with Node.js 12.x

Found bindings for the following environments:
  - Linux 64-bit with Node.js 14.x

This usually happens because your environment has changed since running `npm install`.
Run `npm rebuild node-sass` to download the binding for your current environment.


npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! volunteer-portal@0.5.0 build: `react-scripts build`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the volunteer-portal@0.5.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2020-12-22T14_40_40_311Z-debug.log
```

Calvinxc1 committed:
Working build, missing wait-for-it in api

adamfortuna commented:
> Getting the following when I try to run `docker-compose build` on the new code:

Hrm, that's a weird one. I just tried clearing all of my docker containers and running a `docker-compose build` from scratch and it's not happening. What OS are you using? I'm wondering if this is a Mac vs Windows issue. Reading through [this article](https://jdlm.info/articles/2019/09/06/lessons-building-node-app-docker.html) on Docker makes me think there's going to be some additional fixes we'll need to add to Windows if that's the case in order to map users. Hopefully it's something more simple though.

Calvinxc1 commented:
> > Getting the following when I try to run `docker-compose build` on the new code:
> 
> Hrm, that's a weird one. I just tried clearing all of my docker containers and running a `docker-compose build` from scratch and it's not happening. What OS are you using? I'm wondering if this is a Mac vs Windows issue. Reading through [this article](https://jdlm.info/articles/2019/09/06/lessons-building-node-app-docker.html) on Docker makes me think there's going to be some additional fixes we'll need to add to Windows if that's the case in order to map users. Hopefully it's something more simple though.

Linux ;-)

adamfortuna commented:
> Linux ;-)

Back to the drawing board. 😅

adamfortuna committed:
API, tests and client running together. Not able to verify DB connection yet

Calvinxc1 committed:
Force rebuild of node_modules properly

adamfortuna committed:
Change node modules folder volume

adamfortuna committed:
Docker compose up from scratch with a single command

adamfortuna committed:
Getting the front-end working end to end

adamfortuna committed:
Updated the readme for docker updates

adamfortuna committed:
Move python config.yml into the repo since we can use it on production

adamfortuna committed:
Fixed existing tests and updated db to include the uuid columns

adamfortuna committed:
Fixing tests for initiatives relationships to events and roles

adamfortuna committed:
Added personal donation link requests

adamfortuna committed:
Merged master back into this branch

adamfortuna committed:
Clean out node_modules before build of the client

adamfortuna committed:
Fix donation email reference in api

adamfortuna committed:
Fix API for donation requests

jtuckerk commented on docker-compose.yml:
```diff
@@ -0,0 +1,59 @@
+
+  client:
+    build: ./client
+    command: npm run start
```
Let's do "npm run client"
run start runs from the build directory and requires a rebuild to pull in changes.
"npm run client" will allow changes to be hot reloaded.

When we break out services/compose files for prod vs dev we'll use npm run start in prod there to use the optimized build

adamfortuna replied:
Sounds like a plan, changing this one.

jtuckerk commented on Makefile:
```diff
@@ -0,0 +1,30 @@
+db-reload-dev:
+	docker-compose run --rm api python bootstrap.py
+db-reload-dev:
+	docker-compose run --rm api ENV=test python bootstrap.py
```
This wasn't working for me

adamfortuna replied:
Updated these to use a new test container and updated the makefile.

jtuckerk commented on api/data_sources.py:
```diff
@@ -173,14 +174,47 @@ def create_linked_model_object(self, row: ResultProxy) -> Optional[Any]:
-        return existing_linked_datasets[model]
-    else:
-        return None
+        # iterate over the parameters of the model class
```
I think this whole file can be deleted, but this section in particular looks like a regression to an older commit

adamfortuna replied:
Yeah, I think it can. If we need parts of it, we can move them into the new spots later. Deleting these for now.

jtuckerk commented on api/deploy-prod-app.sh:
```diff
@@ -1,8 +1,8 @@
 
 branch_name=$(git symbolic-ref --short -q HEAD)
-if [ "$branch_name" = "production" ]; then
+if [ "$branch_name" = "master" ]; then
```
this shouldn't be reverted to master

adamfortuna replied:
Good catch!

jtuckerk commented on client/deploy-prod-client.sh:
```diff
@@ -1,10 +1,10 @@
 
 branch_name=$(git symbolic-ref --short -q HEAD)
-if [ "$branch_name" = "production" ]; then
+if [ "$branch_name" = "master" ]; then
```
This shouldn't be reverted to master

adamfortuna replied:
Good catch!

jtuckerk reviewed (changes requested):
I think we're pretty close to being good enough to merge this with develop.

It looks like this PR reverts some changes. I think was happened was a rebase/merge/gitword in  6a4d9bb back onto master. (edit: master is gone, i'm not sure exactly what happened but it was definitely caused by switching from master to develop after this branch was started) the should have been a rebase onto develop

2 PRs to merge to this one:
#76 makes slight changes to the docker configs that I think will same others some frustration
#75 is working, but not thoroughly tested, code for the airtable data pipelines

Some other things that I think we can finish later:

- the data source/sink code is now obsolete, I believe
- editing the db columns will require changes in the models, schemas, pipeline converters, and fake data utils. while this isnt ideal, I dont think it's a blocker.
- linting and type annotations (I need to do this in the pipeline code as well)

adamfortuna committed:
Map personal donation table as part of bootstrapping

adamfortuna committed:
Setup the testing environment as its own docker container based on the API container

adamfortuna committed:
Removing deprecated data_source object in favor of sqlalchemy

adamfortuna committed:
Revert from master back to develop

adamfortuna committed:
Added donation request API endpoint

jtuckerk commented on api/api.py:
```diff
@@ -56,7 +56,7 @@ def get_initiative_by_external_id(initiative_external_id, db: Session = Depends(
-    donationEmail = DonationEmail(email=email)
-    DonationEmail.insert(donationEmail)
+def create_donation_link_request(donationEmail: DonationEmailSchema, db: Session = Depends(get_db)) -> DonationEmailSchema:
+    db.add(DonationEmail(email=donationEmail.email))
```
The old way looked weird. I was surprised it worked. or did it not?

adamfortuna replied:
I haven't been able to validate the previous way works, but I know the new way works at least. I wanted to make it cleaner, but gave up eventually and went the less-pretty way.

jtuckerk commented on api/schemas/donation_email.py:
```diff
@@ -1,13 +1,31 @@
     email: str
-    request_sent: datetime
+
+    @validator('email')
```
Theres a pydantic type we can use to do this validation
`from pydantic import EmailStr`

jtuckerk commented on client/src/App.js:
```diff
@@ -1,11 +1,9 @@
 import React from 'react';
```
I believe this was an unintended revert too, along with all of the other changes in client/src I think

adamfortuna replied:
Off, yeah there shouldn't be any changes in Client other than cleaning up an API request path or two to no longer include a trailing slash. Must've had conflicts during a merage. I'll replace that.

jtuckerk reviewed (changes requested):
I think there are still a bunch of changes in the client that revert a fair bit of work in development

should there be any changes in this PR to the files in client/src? 

The rest of the changes look good tho!

jtuckerk commented on docker-compose.yml:
```diff
@@ -0,0 +1,59 @@
+  proxy:
+    image: nginx
+    depends_on:
+    - api
```
can we add api-test here? IIUC this only affects the spin-up order but wont spin up both api and api-test at the same time when the proxy reloads

jtuckerk replied:
Nvm, with the PR i have  out this may be obsolete

jtuckerk commented:
changes from #75, #78 and #80 all merged into and on top of this and appear to be working in branch revert+docker+db_pipeline

adamfortuna committed:
Switch donation email to use email pydantic type

adamfortuna committed:
Merge branch 'develop' into adam-multiple-environments

adamfortuna committed:
Mering develop/client back in

adamfortuna committed:
Update GitHub tests to run python using new bootstrap setup

adamfortuna committed:
Switching to a single CI file that uses docker compose for github testing

adamfortuna committed:
Fix client CI command for GH workflow

adamfortuna committed:
Fixing the client docker name to the actual name and not the makefile alias

adamfortuna committed:
Install all depenencies with NPM for CI purposes

adamfortuna committed:
Exit client tests in CI after running

adamfortuna committed:
Fix spacing in github workflow file to be consistent

jtuckerk committed:
Put back client code to latest develop changes. (#80)

* Put back client code to latest develop changes.

* add missing paren

* fix initiatives path
